### PR TITLE
Adjust samplesPerChunk from 120 to 220

### PR DESCRIPTION
**Pull request by leizor**
This PR suggests a new value for `samplesPerChunk` based on some observations/assumptions about the average dataset:

- \>95% of timestamps have uniform periods, resulting in a dod of 0.
- ~60% of values are the same as the previous value, resulting in an xor of 0.
- ~40% of values can re-use the same “interesting bits” as the previous value, so we can use the `10` control bits.
- The average length of interesting bits of a value is ~8 bits.
- 99.9% of the first samples of a chunk used a 2-byte uvarint to encode the ts delta (16 bits).

Therefore, this equation (n = number of samples) for expected size per chunk covers 60% of the cases:

```
S = (n-1) * (1+1)  // Size of samples after the first sample.
+ 16 + 64          // Size of first sample (2 bytes for ts delta + 64 bytes for float val).
+ 64               // 64 bit header w/ timestamp.

= 2n + 142
```

The other 40% is covered by this:

```
S = (n-1) * (1 + (2+8))  // Size of samples after first sample (1 bit for ts dod, 2 control 
                         // bits + 8 bit average xor length for value)
+ 16 + 64 + 64

= 11n + 133
```

Combining the two cases covers nearly 100% of the cases:

```
// Size after compression as a function of # of samples (n)
Sc = (2n + 142) * .6 + (11n + 133) * .4

= 5.6n + 138.4
```

This allows us to arrive at an equation to describe the expected compression ratio:

```
CR = 128n / (5.6n + 138.4)
```

That means that the compression ratio approaches an asymptote of 128 / 5.6 = 22.86. If we target achieving 90% of the maximum expected compression ratio, we'd be targeting 22.86 * .9 = 20.574. Solving for n, we arrive at 223 samples per chunk, which is 1388 bits or 174 bytes.

As a point of comparison, the current samples per chunk value yields ~83% of the maximum expected compression ratio with an expected chunk size of 811 bits or 102 bytes.

### CODE REVIEW

**Comment by bboreham** [comment]
It appears to use slightly more memory: 
image

and slightly less disk space:
image

There are fewer chunks in memory, as expected:
image

@leizor what metric(s) were you looking to show an improvement on?

(Also it seems Prombench is executing no queries.  That seems like a crucial omission)

**Comment by leizor** [comment]
I think saving a bit of disk space in exchange for using slightly more memory is pretty much what I expected. 

It's a bummer about the queries though because I suspect that this change might also result in slightly longer query latencies since each chunk is larger (I was hoping it'd not be the case though). Is it something in my PR that's making Prombench not execute any queries?

**Comment by leizor** [comment]
Converting this PR to a draft; I'm going to take a stab at forgoing sample count limits in favor of chunk size limits [here](https://github.com/prometheus/prometheus/pull/12054) instead.

**Comment by beorn7** [comment]
Any ideas why it uses more RAM? With fewer chunks and a better compression ratio, I would expect _less_ RAM usage. If at all, I would expect more CPU usage and longer query times (because we have to iterate through chunks with more samples).

**Comment by leizor** [comment]
> Any ideas why it uses more RAM? With fewer chunks and a better compression ratio, I would expect less RAM usage. If at all, I would expect more CPU usage and longer query times (because we have to iterate through chunks with more samples).

Hmm good point. I was thinking it'd use more RAM because it'd be holding more samples in memory at any given time, but I think you're right--the number of samples should actually still be the same, just in fewer chunks.

**Comment by bboreham** [comment]
Good point. 

Small note: if the data is queried it will be brought back into memory. So it’s relevant that Prombench was not running any queries. 
(However it’s likely that data collected will be much more than data queried, so this is not a huge effect)

**Comment by bwplotka** [comment]
Side note: If we are considering changing that number - what about making this (e.g. hidden for now) flag?

**Comment by beorn7** [comment]
Yeah, sure, if we end up with tweaking the number of samples per chunk, it could be a hidden flag for the time being.

But maybe going down the path of trying to set a target in _bytes_ rather than _samples_ is anyway better. (Which then, of course, might be a (hidden) flag, too.)

**Comment by leizor** [comment]
Do we want to add the flag in this PR or defer to later one?

**Comment by beorn7** [comment]
I would say, let's put it into a (hidden) flag here in this PR (similar to `--storage.tsdb.min-block-duration` and friends, e.g. `--storage.tsdb.max-samples-per-chunk`).

With the flags, users can experiment, and based on the experiences, we can then set a new default limit that's generally better (or maybe we go down the path of limiting the number of bytes after all).

But if we hardcode a different limit now, and then it turns out it's detrimental in some cases, it will create a needless confusion.

**Comment by bwplotka** [comment]
I would say we need a flag.

**Comment by leizor** [comment]
Ok, I've added the new flag `storage.tsdb.samples-per-chunk` and defaulted it to the original value of 120 (I think that's what we landed on).

**Review by jesusvazquez — APPROVED** [review]
LGTM, nice job @leizor 

I am also uncertain about why memory slightly increased with 220 samples, would be good to look into that.

Also good call on adding the flag.

**Comment by beorn7** [comment]
Thanks, merging now…

**Comment by colega** [comment]
I've got two notes about this:

One is regarding the memory usage: for which I guess one of the reasons are the 8 extra bytes of that setting we're storing per memSeries. 

What was the reason to do this setting per memSeries? Is it to make sure that the 1/4th chunk estimation works correctly if the setting is changed in the Head? But this doesn't sound like a setting someone would want to change in runtime, and even if they did, the side-effect of having uneven chunks once doesn't seem to be an argument to store that integer in every single series.

The other one is regarding how that _strange_ amount of 220 samples (which isn't a multiple of 15, 30, 60: the usual Prometheus scraping intervals) really works with the chunk size estimation:

https://github.com/prometheus/prometheus/blob/37e5249e33e4c2d7bc49c4ac587faf12eb6785b7/tsdb/head_append.go#L1371-L1377

I was wondering if [the estimation algorithm](https://github.com/prometheus/prometheus/blob/37e5249e33e4c2d7bc49c4ac587faf12eb6785b7/tsdb/head_append.go#L1391-L1402) would maybe try to issue 3 x 160-samples chunks for a 15s scrape interval (respecting the 220 maximum, making it useless)? But after writing some tests I see that setting this at 220, we'll actually issue just 240-samples chunks:

```diff
diff --git tsdb/head_test.go tsdb/head_test.go
index af3df378e..7b69f11c1 100644
--- tsdb/head_test.go
+++ tsdb/head_test.go
@@ -1318,6 +1318,20 @@ func TestComputeChunkEndTime(t *testing.T) {
                        max:   1000,
                        res:   104,
                },
+               {
+                       // Test 15s scrape inteval with 220 samples per chunk.
+                       start: 0,                // Start at 0 to make it easier.
+                       cur:   15e3 * (220 / 4), // We're at 1/4th after receiving 220/4 samples each 15s.
+                       max:   2 * 60 * 60e3,    // 2h
+                       res:   1 * 60 * 60e3,    // 1h, i.e. 2-240 samples chunks per hour
+               },
+               {
+                       // Test 30s scrape inteval with 220 samples per chunk.
+                       start: 0,                // Start at 0 to make it easier.
+                       cur:   30e3 * (220 / 4), // We're at 1/4th after receiving 220/4 samples each 30s.
+                       max:   2 * 60 * 60e3,    // 2h
+                       res:   2 * 60 * 60e3,    // 4, i.e. 1 240-samples chunk.
+               },
        }
 
        for _, c := range cases {
```

Maybe we should change the setting to 240 then? I know this is specific to the default 2h blocks, but I don't think anyone runs 110-minutes blocks, and it's strange to see chunks issued with a different size compared to what was set.

**Comment by alanprot** [comment]
Should we only set the config if its `!= 0` here?

Just got a divided by 0 error when updating prometheus on thanos! :D

https://github.com/thanos-io/thanos/pull/6392

https://github.com/prometheus/prometheus/blob/37e5249e33e4c2d7bc49c4ac587faf12eb6785b7/tsdb/db.go#L790

**Comment by leizor** [comment]
@colega 

> What was the reason to do this setting per memSeries?

It was just a way to pass the information down to where it was needed; we need to know what the value of samplesPerChunk is in `appendPreprocessor` of memSeries but memSeries doesn't have a reference to Head or anything to get that information otherwise.

I think I'm going to give memSeries a reference to HeadOptions instead. Unfortunately it'll use the same extra 8 bytes per memSeries but I figure at least it wouldn't imply that the value could be different between memSeries.

**Comment by beorn7** [comment]
Updated the PR title to not cause confusion in the future.

File: tsdb/head.go [hunk 1]
```diff
@@ -1915,7 +1920,8 @@ type memSeries struct {
 
 	mmMaxTime int64 // Max time of any mmapped chunk, only used during WAL replay.
 
-	nextAt int64 // Timestamp at which to cut the next chunk.
+	samplesPerChunk int   // Target number of samples per chunk.
```
Comment (codesome) [inline tsdb/head.go:1923]: We should probably make this `uint16` and have validation during init to keep it within that range. Saves 48 bytes per series.

Comment (leizor) [reply tsdb/head.go:1923]: We ended up pulling that out in https://github.com/prometheus/prometheus/pull/12390.

**Comment by codesome** [comment]
I am wondering if setting it to a higher number gets hit by https://github.com/prometheus/prometheus/pull/8582. I see that the logic still exists.

**Pull request merged**